src/main/java/uk/gov/ida/verifyserviceprovider/services/UnsignedAssertionsResponseHandler.java

```diff
@@ -0,0 +1,129 @@
+package uk.gov.ida.verifyserviceprovider.services;
+
+import org.opensaml.saml.saml2.core.Assertion;
+import org.opensaml.saml.saml2.core.Attribute;
+import org.opensaml.saml.saml2.core.Response;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import uk.gov.ida.saml.core.IdaConstants;
+import uk.gov.ida.saml.core.extensions.eidas.CountrySamlResponse;
+import uk.gov.ida.saml.core.extensions.eidas.EncryptedAssertionKeys;
+import uk.gov.ida.saml.core.validation.SamlResponseValidationException;
+import uk.gov.ida.saml.deserializers.StringToOpenSamlObjectTransformer;
+import uk.gov.ida.saml.security.AssertionDecrypter;
+import uk.gov.ida.saml.security.EidasValidatorFactory;
+import uk.gov.ida.saml.security.SecretKeyDecryptorFactory;
+import uk.gov.ida.saml.security.exception.SamlFailedToDecryptException;
+import uk.gov.ida.saml.security.validators.ValidatedResponse;
+import uk.gov.ida.saml.security.validators.encryptedelementtype.EncryptionAlgorithmValidator;
+import uk.gov.ida.verifyserviceprovider.validators.InstantValidator;
+
+import javax.crypto.NoSuchPaddingException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import static uk.gov.ida.saml.security.errors.SamlTransformationErrorFactory.unableToDecrypt;
+
+public class UnsignedAssertionsResponseHandler {
+    private static final Logger LOG = LoggerFactory.getLogger(UnsignedAssertionsResponseHandler.class);
+
+    private final EidasValidatorFactory eidasValidatorFactory;
+    private final StringToOpenSamlObjectTransformer<Response> stringToResponseTransformer;
+    private final InstantValidator instantValidator;
+    private final SecretKeyDecryptorFactory secretKeyDecryptorFactory;
+    private final EncryptionAlgorithmValidator encryptionAlgorithmValidator;
+
+    public UnsignedAssertionsResponseHandler (
+            EidasValidatorFactory eidasValidatorFactory,
+            StringToOpenSamlObjectTransformer<Response> stringToResponseTransformer,
+            InstantValidator instantValidator,
+            SecretKeyDecryptorFactory secretKeyDecryptorFactory,
+            EncryptionAlgorithmValidator encryptionAlgorithmValidator
+    ) {
+        this.eidasValidatorFactory = eidasValidatorFactory;
+        this.stringToResponseTransformer = stringToResponseTransformer;
+        this.instantValidator = instantValidator;
+        this.secretKeyDecryptorFactory = secretKeyDecryptorFactory;
+        this.encryptionAlgorithmValidator = encryptionAlgorithmValidator;
+    }
+
+    public ValidatedResponse getValidatedResponse(
+            Assertion hubResponseAssertion,
+            String expectedInResponseTo
+    ) {
+        ValidatedResponse validatedResponse = eidasValidatorFactory.getValidatedResponse(
```

Comment:
It's not a part of this review, as you didn't write it, but `eidasValidatorFactory.getValidatedResponse` has quite a few possible interpretations in my head:

- validate the response (best guess)
- call out to another service to validate the response (similar, but more misleading)
- extract a validated response from inside the passed in parameter
- represent the passed in parameter as validated (ie. assumption is it's already validated, so just cast/copy/box it into a ValidatedResponse, a method you might call `asValidatedResponse`)

I took a look in `saml-lib` and `getValidatedResponse` seems to be doing a validation on the signature(s) in the response using a `SamlResponseSignatureValidator`. My instinct would be to rename that method to `validateResponseSignature`, as it's an action not a getter.

I know it's not for now, and for some reason my slight confusion has led to a small essay, but hopefully it helps to make clear what I mean when I say the context is missing from some things - which means it takes more time to understand what they're doing, and is difficult to remember.

No action for this review, but perhaps we can track a list of things we could do to make things a bit clearer for future developers. (I'm a little worried too that as it's in `saml-lib`, changing the name of the method will lead to issues for a number of other projects.)

Reply:
It could definitely be named better, but I think you're right that it's used across different projects and so would be hard to change.